### PR TITLE
feat(forge): change startPrank to overwrite existing prank instead of erroring

### PR DESCRIPTION
**Pull request by 4meta5**
Closes https://github.com/foundry-rs/foundry/issues/4779

## Motivation

There is user demand for a `startPrank` or `changePrank` cheatcode that allows optimistically setting a new prank such that any existing prank is overwritten and there is no error if the existing prank is not set.

## Solution

`startPrank` is changed to allow overwriting existing pranks instead of erroring. It is not able to overwrite existing pranks unless the existing prank has already been applied at least once.

### CODE REVIEW

**Review by mattsse — APPROVED** [review]
codewise lgtm,

wdyt @mds1

**Comment by mds1** [comment]
Should we just change the behavior of `startPrank` to work this way instead of introducing a new cheat? It feels intuitive that `startPrank` would end any active pranks in order to start a new one. Then we'd also want a PR to forge-std that updates `changePrank` to just revert explaining this breaking change. cc @PaulRBerg 

Also it looks like this only adds `startChangePrank(address,address)`, but we also need `startChangePrank(address)` (or, if we just change `startPrank`, make sure to change behavior in both

Also one edge case to make sure we handle is:

```solidity
vm.startPrank(msgSender, txOrigin);
// --- do stuff ---

// now we change prank, but need to make sure to reset txOrigin
vm.startPrank(msgSender); 
```

**Comment by PaulRBerg** [comment]
> Should we just change the behavior of startPrank to work this way

Works for me

**Review by mattsse — APPROVED** [review]
lgtm,

pending @mds1 and lint

**Comment by mds1** [comment]
Looks good! Just two questions:
1. What if I call `vm.prank` immediately followed by `vm.prank` (same question if either/both of those are replaced with `startPrank`)? I think we should make sure that reverts, because the first prank was never used
2. I don't see new tests for `startPrank`, just tests for `prank`, so just want to confirm the above question, which was:

> Also one edge case to make sure we handle is:
> 
> ```solidity
> vm.startPrank(msgSender, txOrigin);
> // --- do stuff ---
> 
> // now we change prank, but need to make sure to reset txOrigin
> vm.startPrank(msgSender); 
> ```

**Comment by 4meta5** [comment]
Both suggestions are reasonable.

> 1. What if I call `vm.prank` immediately followed by `vm.prank` (same question if either/both of those are replaced with `startPrank`)? I think we should make sure that reverts, because the first prank was never used

Yes, will add to this PR.

> 2. I don't see new tests for `startPrank`, just tests for `prank`, so just want to confirm the above question, which was:
> 
> > Also one edge case to make sure we handle is:
> > ```solidity
> > vm.startPrank(msgSender, txOrigin);
> > // --- do stuff ---
> > 
> > // now we change prank, but need to make sure to reset txOrigin
> > vm.startPrank(msgSender); 
> > ```

Will add tests for `startPrank` as well

**Review by mattsse — APPROVED** [review]
lgtm, this seems to work!

**Comment by 4meta5** [comment]
How do we use a prank in the tests? Specifically, I want to invoke Cheatcodes::call such that the prank is applied here https://github.com/foundry-rs/foundry/pull/4826/files#diff-3a6637039aa2043110b6bf828cf9c802d7621bdd472a536ea43f56421026eb64R597-R616 ? This would require executing a call that that uses the prank. I don't see any example of this in `Prank.t.sol`.

**Comment by Evalir** [comment]
hey hey @4meta5 thanks for the work! I've pushed some commits to your branch to get it over the line, hope you don't mind:
- rebased to make sure it's up to par with master
- added the `startPrank` tests and tweaked an error being returned to use one of our internal utils.

cc @mattsse / @mds1 should be g2g bar CI (github is having problems)

**Comment by mds1** [comment]
@Evalir just confirming if you or @4meta5 handled these items too / have test cases for them? https://github.com/foundry-rs/foundry/pull/4826#issuecomment-1527749789

**Comment by Evalir** [comment]
Ah whoops, seems I stashed & didn't upload that edge case: It is handled, but just pushed the missing test

File: testdata/cheats/Prank.t.sol [hunk 1]
```diff
@@ -170,6 +170,28 @@ contract PrankTest is DSTest {
...
+        victim.assertCallerAndOrigin(
+            sender, "msg.sender was not set during prank", oldOrigin, "tx.origin invariant failed"
+        );
+
+        cheats.stopPrank();
+        // Ensure we cleaned up correctly after stopping the prank
+        victim.assertCallerAndOrigin(
+            address(this), "msg.sender was not cleaned up", oldOrigin, "tx.origin invariant failed"
+        );
+    }
```
Comment (Evalir) [inline testdata/cheats/Prank.t.sol:193]: Just to confirm: this should be the edge case you're talking about @mds1 right?

Comment (mds1) [reply testdata/cheats/Prank.t.sol:193]: Yep that's it, thanks!

**Review by mattsse — APPROVED** [review]
sweet,

tysm @4meta5

**Comment by PaulRBerg** [comment]
This is awesome! I'm glad that this was implemented.

Am I correct that now I can replace all of our uses of [`changePrank`](https://github.com/foundry-rs/forge-std/blob/7f7cb06aa55fc75d2280d7d54713dc4368d71fd3/src/StdCheats.sol#L545) with `vm.startPrank`?

**Comment by urataps** [comment]
Cool feature, but should contract deployment count also as if state was applied?

Currently I have some code that fails with this:

```
vm.startPrank(owner);
// deploy some contracts

vm.startPrank(stranger)
// check some stuff
```

**Pull request merged**